### PR TITLE
FE-487 BLE-centralization of code and code optimizations

### PR DESCRIPTION
**Pull request by PavlosTze**
## **Why?**

In multiple view models the functions of scan, connect, pair, disconnect share the same code more or less. Explore potential centralization of that code in order to have a central source of truth.

### **How?**
1. Create `open class BluetoothHeliumViewModel` which acts as a base class for Claim, Reboot, Change Frequency, OTA of Helium `ViewModels` respectively.
2. That base class is responsible for scanning internally and connecting to a specific station and exposes the **open** functions `onPaired`, `onNotPaired`, `onScanFailure`, `onConnected` and `onConnectionFailure` which will be overridden by the view models in order to use them respectively.
3. Reboot, Change Frequency and OTA ViewModels use the full functionality of `BluetoothHeliumViewModel` for their own purposes and only deal with UI-related stuff based on the result, whereas all Claim-related ViewModels of Helium do the scanning on their own and use only the "Connection" part of the `BluetoothHeliumViewModel`.

### **Testing**

Test all BLE-related functionalities, try with different combinations (paired vs non-paired):
1. Claiming of Helium (we want BLE-related to be OK, and get an error at the end that device is already claimed)
4. Reboot of Helium Station (through station settings)
5. Device OTA of Helium Station (through station settings, in order to enable it set the `if` as `true` in `DeviceSettingsViewModel.kt` lines 223)
6. Change Frequency of Helium Station (through station settings)

### CODE REVIEW

**Review by JohnFrontzos — CHANGES_REQUESTED** [review]
Well structured Pavlos! I have commented some minor changes only.  
Also, would prefer something else for the "need to overridden" open fun, but I don't have any other solution at the moment.

File: app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt [hunk 1]
```diff
@@ -20,10 +19,10 @@ import timber.log.Timber
 
 class ClaimHeliumPairViewModel(
     private val resources: Resources,
-    private val analytics: Analytics,
+    analytics: Analytics,
     private val scanDevicesUseCase: BluetoothScannerUseCase,
-    private val connectionUseCase: BluetoothConnectionUseCase
-) : ViewModel() {
+    connectionUseCase: BluetoothConnectionUseCase
+) : BluetoothHeliumViewModel("", null, connectionUseCase, analytics) {
```
[the remark this answers is not in the document]
Comment (PavlosTze) [reply app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt:25]: @JohnFrontzos Not sure what you mean, this doesn't work. Maybe you meant to define a `const val EMPTY_STRING = ""` ?

File: app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt [hunk 2]
```diff
@@ -64,57 +62,63 @@ class ClaimHeliumPairViewModel(
...
+
+        super.stopScanning()
+        super.scannedDevice = scannedDevice
+
+        viewModelScope.launch {
+            super.setPeripheralAndConnect(true)
+        }
+    }
+
     init {
```
[the remark this answers is not in the document]
Comment (PavlosTze) [reply app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt:122]: There is no problem regarding the two `init`, they are different, the one in the Base class checks for the pairing status, which is only invoked after the scanning is completed (which the `init` in that ViewModel deals with).

File: app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (JohnFrontzos) [reply app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt:25]: I mean something like fun String.empty() = "", in a single file and would be reused from there

[the remark this answers is not in the document]
Comment (PavlosTze) [reply app/src/main/java/com/weatherxm/ui/claimdevice/helium/pair/ClaimHeliumPairViewModel.kt:25]: I already checked it, that doesn't work in our case. We don't have a string beforehand to do that

**Review by pantelisss — APPROVED** [review]
Codewise seems ok, could not tested on real device

**Review by JohnFrontzos — APPROVED** [review]

**Pull request merged**